presto-native-execution/presto_cpp/main/PrestoExchangeSource.h

```diff
@@ -14,13 +14,50 @@
 #pragma once
 
 #include <folly/Uri.h>
+#include <folly/futures/Retrying.h>
 
 #include "presto_cpp/main/common/Configs.h"
 #include "presto_cpp/main/http/HttpClient.h"
 #include "velox/common/memory/Memory.h"
 #include "velox/exec/Exchange.h"
 
 namespace facebook::presto {
+
+class Backoff {
+ public:
+  Backoff(int64_t maxWaitMs = 1000)
+      : maxWaitMs_(maxWaitMs), startMs_(velox::getCurrentTimeMs()) {}
+
+  // Returns the delay in millis to wait before next try. This is an exponential
+  // backoff delay with jitter. The first call to this always returns 0.
+  int64_t nextDelayMs() {
+    ++numTries_;
+    if (numTries_ == 1) {
+      return 0;
+    }
+    auto rng = folly::ThreadLocalPRNG();
+    static constexpr int64_t kMinBackoffMs = 100;
+    static constexpr int64_t kMaxBackoffMs = 10000;
+    static constexpr double kJitterParam = 0.1;
+    return folly::futures::detail::retryingJitteredExponentialBackoffDur(
+               numTries_ - 1,
+               std::chrono::milliseconds(kMinBackoffMs),
+               std::chrono::milliseconds(kMaxBackoffMs),
+               kJitterParam,
+               rng)
+        .count();
+  }
+
+  bool isExhausted() const {
```

Comment:
Noob question: Is there any reason to get the exhaust signal based on tries instead of time? For example: we could say skip retry after 20 attempts? If both are same, going with tries is efficient because its just a simple comparison with a number and no need to get current time from a system call.

Reply:
I believe we should just do what Java does here. 

This is a tradeoff between reliability(larger value is better), and failure latency(smaller value is better). Tracking error duration gives direct visibility into it.

There is no question of efficiency because get time is super fast, and takes no context switch.

Reply:
I found this SO answer which gives some visibility on how this is implemented, and is just a user-space memory read: https://unix.stackexchange.com/questions/316074/why-dont-linux-utils-use-a-system-call-to-get-the-current-time

Reply:
Can you comment when isExhausted condition met? Thanks!